spread/client.go

```diff
@@ -281,26 +281,32 @@ func (c *Client) run(script string, dir string, env *Environment, mode outputMod
 		rebootKey = rerr.Key
 		output = append(output, '\n')
 
-		timedout := time.After(c.killTimeout)
-		err := c.Run(fmt.Sprintf("reboot &\nsleep %.0f", c.killTimeout.Seconds()), "", nil)
+		uptime, err := c.getUptime()
 		if err != nil {
-			err = c.Run("echo should-have-disconnected", "", nil)
+			return nil, fmt.Errorf("failed to get uptime before reboot, error: %v", err)
 		}
-		if err == nil {
-			select {
-			case <-timedout:
-				return nil, fmt.Errorf("kill-timeout reached on %s while waiting for reboot", c.job)
-			default:
-			}
-			return nil, fmt.Errorf("reboot request on %s failed", c.job)
-		}
-		if err := c.dialOnReboot(); err != nil {
+		c.Run(fmt.Sprintf("reboot"), "", nil)
+		if err := c.dialOnReboot(uptime); err != nil {
 			return nil, err
 		}
 	}
 	panic("unreachable")
 }
 
+func (c *Client) getUptime() (time.Time, error) {
+	uptime, err := c.Output("date -u --iso-8601=seconds -d \"`cut -f1 -d. /proc/uptime` seconds ago\"", "", nil)
+	if err != nil {
+		return time.Now(), err
+	}
+
+	// Convert unsupported +0000 to +00:00
+	parts := strings.Split(string(uptime), "+")
+	if (len(parts) == 2) && (! strings.Contains(parts[1], ":")) {
+		parts[1] = "00:00"
+	}
+	return time.Parse(time.RFC3339, strings.Join(parts, "+"))
```

Comment:
```
if err != nil {
    return time.Time{}, fmt.Errorf("cannot parse the remote system uptime: %q", uptime)
```
Note that in this case the value after the colon is the actual time, not the error. That's what we need to tell what went wrong.